### PR TITLE
cl_khr_command_buffer_mutable_dispatch + minor fixes

### PR DESCRIPTION
**Pull request by franz**

### CODE REVIEW

File: lib/llvmopencl/LLVMUtils.h [hunk 1]
```diff
@@ -58,9 +58,14 @@ typedef std::map<llvm::Function*, llvm::Function*> FunctionMapping;
 constexpr unsigned NumWorkgroupVariables = 21;
 extern const char *WorkgroupVariablesArray[];
 extern const std::vector<std::string> WorkgroupVariablesVector;
+// work-item function names
 constexpr unsigned NumWIFuncNames = 11;
 extern const char *WIFuncNameArray[];
 extern const std::vector<std::string> WIFuncNameVec;
+// functions that should not be inlined before workgroup pass: WI funcs + printf funcs
```
Comment (pjaaskel) [inline lib/llvmopencl/LLVMUtils.h:65]: Can you expand a bit for documentation: Why should not they be inlined?

File: lib/CL/devices/common.c [hunk 2]
```diff
@@ -213,7 +213,7 @@ llvm_codegen (char *output, unsigned device_i, cl_kernel kernel,
   const char *cmd_line[64]
     = { pocl_get_path ("CLANG", CLANG), "-o", tmp_module, tmp_objfile };
   unsigned last_arg_idx = 4;
-  /* immediate flush enabled results in "__printf_flush_buffer" symbol
+  /* immediate flush enabled results in "pocl_flush_printf_buffer" symbol
    * referenced it the built kernel.so, however that function exists only
```
Comment (pjaaskel) [inline lib/CL/devices/common.c:217]: *in* the built

File: lib/kernel/printf.c [hunk 3]
```diff
@@ -67,18 +67,18 @@ __pocl_printf_alloc (PRINTF_BUFFER_AS char *__buffer,
...
- * this is a stub that will be replaced by __pocl_printf_alloc in Workgroup.cc
+ * this is a stub that will be replaced by pocl_printf_alloc in Workgroup.cc
  */
 PRINTF_BUFFER_AS void *ATTRS
-__printf_alloc (uint32_t bytes)
+pocl_printf_alloc_stub (uint32_t bytes)
 {
-  /* this ensures __pocl_printf_alloc is not optimized away */
-  PRINTF_BUFFER_AS void *retval = __pocl_printf_alloc (
+  /* this ensures pocl_printf_alloc is not optimized away */
```
Comment (pjaaskel) [inline lib/kernel/printf.c:75]: Could we use the "aux functions" for this instead?

File: lib/llvmopencl/Flatten.cc [hunk 4]
```diff
@@ -74,9 +74,9 @@ static bool flattenAll(Module &M) {
 
   for (llvm::Module::iterator i = M.begin(), e = M.end(); i != e; ++i) {
     llvm::Function *F = &*i;
-    if (F->isDeclaration() || F->getName().starts_with("__pocl_print") ||
-        F->getName() == "__printf_alloc" ||
-        F->getName() == "__printf_flush_buffer" ||
+    if (F->isDeclaration() || F->getName() == "pocl_printf_alloc" ||
+        F->getName() == "pocl_printf_alloc_stub" ||
```
Comment (pjaaskel) [inline lib/llvmopencl/Flatten.cc:78]: Why not add these to AuxFuncs for the targets that require the host-side printf functionality?

File: README.md [hunk 5]
```diff
@@ -28,37 +28,78 @@ tools:
...
+
+### Mac OS X
+
+| CPU device  |     LLVM 16    |     LLVM 17     |
+|:------------|:--------------:|:---------------:|
+| [Apple M1](https://github.com/pocl/pocl/actions/workflows/build_macos.yml) | :green_circle: | :green_circle:  |
+
+### Windows
+
+| CPU device  |     LLVM-MinGW 18    |  LLVM-MinGW 19  |
```
Comment (pjaaskel) [inline README.md:99]: MSVC?

File: include/CL/cl_ext.h [hunk 6]
```diff
@@ -62,7 +62,8 @@ typedef cl_uint             cl_command_buffer_state_khr;
 typedef cl_properties       cl_command_buffer_properties_khr;
 typedef cl_bitfield         cl_command_buffer_flags_khr;
 typedef cl_properties       cl_command_properties_khr;
-typedef struct _cl_mutable_command_khr* cl_mutable_command_khr;
+/* PoCL change: typedef'd here to pointer to a command node */
```
Comment (pjaaskel) [inline include/CL/cl_ext.h:65]: to *point*

File: include/pocl.h [hunk 7]
```diff
@@ -523,14 +522,17 @@ struct _cl_command_node
...
+  /* true if the node belongs to a cl_command_buffer_khr */
+  cl_int buffered;
 
   /* Which of the command queues in the command buffer's queue list
    * this command was recorded for. */
   cl_uint queue_idx;
   /* List of buffers this command accesses, used for inserting migrations */
   pocl_buffer_migration_info *migr_infos;
+
+  /* pointer to command buffer or NULL */
```
Comment (pjaaskel) [inline include/pocl.h:534]: Pls instead describe when non-NULL (when added to a command buffer - implying there can be only one per our command struct).

File: lib/CL/clUpdateMutableCommandsKHR.c [hunk 8]
```diff
@@ -0,0 +1,165 @@
...
+
+      cl_uint program_dev_i = cfg->command->program_device_i;
+      cl_kernel kernel = cfg->command->command.run.kernel;
+      cl_device_id realdev = pocl_real_dev (dev);
+      cl_uint work_dim = cfg->work_dim
+          ? cfg->work_dim
+          : cfg->command->command.run.pc.work_dim;
+      unsigned copy_size = sizeof (size_t) * work_dim;
+
+      // PoCL does not store Global worksize, only the number of WGs.
```
Comment (pjaaskel) [inline lib/CL/clUpdateMutableCommandsKHR.c:103]: /* */

File: lib/CL/devices/common_utils.c [hunk 9]
```diff
@@ -291,6 +291,17 @@ pocl_cpu_init_common (cl_device_id device)
   device->local_mem_size = pocl_get_int_option ("POCL_CPU_LOCAL_MEM_SIZE",
                                                 device->local_mem_size);
 
+  device->cmdbuf_capabilities
+    = CL_COMMAND_BUFFER_CAPABILITY_SIMULTANEOUS_USE_KHR
+      | CL_COMMAND_BUFFER_CAPABILITY_KERNEL_PRINTF_KHR
+      | CL_COMMAND_BUFFER_CAPABILITY_OUT_OF_ORDER_KHR
+      | CL_COMMAND_BUFFER_CAPABILITY_MULTIPLE_QUEUE_KHR;
+  device->cmdbuf_required_properties = 0;
+  /* Local size (and offset) is not so simple to change,
```
Comment (pjaaskel) [inline lib/CL/devices/common_utils.c:300]: Yet we still support it now? :)

File: lib/CL/devices/pthread/pthread_scheduler.c [hunk 10]
```diff
@@ -290,14 +289,19 @@ work_group_scheduler (kernel_run_command *k,
...
-   * a compilation option), but we need to reset rounding mode after every
-   * iteration (since it can be changed during kernel execution). */
+   * a compilation option) */
   unsigned flush = k->kernel->program->flush_denorms;
-  if (thread_data->current_ftz != flush)
-    {
-      pocl_set_ftz (flush);
-      thread_data->current_ftz = flush;
-    }
+  cl_device_fp_config supports_any_denorms
```
Comment (pjaaskel) [inline lib/CL/devices/pthread/pthread_scheduler.c:294]: Can we refactor the FP setup to a function which both basic and pthread can use?

File: lib/CL/pocl_cl.h [hunk 11]
```diff
@@ -753,8 +753,10 @@ struct pocl_device_ops {
   /** Optional: If the driver uses the default build_poclbinary implementation
    * from common_driver.c, that implementation calls this to compile a
    * "dynamic WG size" kernel. */
-  void (*compile_kernel) (_cl_command_node *cmd, cl_kernel kernel,
-                          cl_device_id device, int specialize);
+  int (*compile_kernel) (_cl_command_node *cmd,
```
Comment (pjaaskel) [inline lib/CL/pocl_cl.h:756]: is prototypes.inc updated?

File: lib/CL/pocl_llvm_build.cc [hunk 12]
```diff
@@ -842,7 +843,8 @@ int pocl_llvm_link_program(cl_program program, unsigned device_i,
   PoclCompilerMutexGuard lockHolder(&llvm_ctx->Lock);
 
   llvm::Module *libmodule = getKernelLibrary(device, llvm_ctx);
-  assert(libmodule != NULL);
+  if (libmodule == nullptr)
```
Comment (pjaaskel) [inline lib/CL/pocl_llvm_build.cc:846]: Perhaps rename to LibraryModule

File: lib/CL/pocl_llvm_build.cc [hunk 13]
```diff
@@ -993,26 +995,23 @@ static llvm::Module *getKernelLibrary(cl_device_id device,
...
-          POCL_MSG_WARN("Using fallback %s as the built-in lib.\n",
-                        kernellib_fallback.c_str());
-          lib = parseModuleIR(kernellib_fallback.c_str(), llvmContext);
-        }
-      else
-        POCL_ABORT("Kernel library file %s doesn't exist.\n", kernellib.c_str());
-    }
-  assert (lib != NULL);
-  kernelLibraryMap->insert(std::make_pair(device, lib));
+  if (pocl_exists(kernellib.c_str())) {
```
Comment (pjaaskel) [inline lib/CL/pocl_llvm_build.cc:1000]: Perhaps rename kernellib to BuiltinLibrary

File: lib/CL/pocl_llvm_build.cc [hunk 14]
```diff
@@ -993,26 +995,23 @@ static llvm::Module *getKernelLibrary(cl_device_id device,
     kernellib_fallback += ".bc";
   }
 
-  llvm::Module *lib;
+  llvm::Module *lib = nullptr;
```
Comment (pjaaskel) [inline lib/CL/pocl_llvm_build.cc:998]: Perhaps rename lib to BuiltinLib

File: lib/CL/pocl_util.h [hunk 15]
```diff
@@ -305,20 +305,32 @@ void pocl_update_event_complete (const char *func, unsigned line,
                                  cl_event event, const char *msg);
 
 POCL_EXPORT
-int pocl_copy_command_node (_cl_command_node *dst_node,
-                            _cl_command_node *src_node);
+void pocl_update_event_failed (const char *func,
```
Comment (pjaaskel) [inline lib/CL/pocl_util.h:308]: While at it, perhaps add a comment when this function should be called.

**Review by pjaaskel — CHANGES_REQUESTED** [review]
Nice stuff! I found nothing major, but perhaps worth checking before pulling in.

File: README.md (hunk 5 above)
Comment (franz) [reply README.md:99]: I have no idea about MSVC ? when Henry finishes his work he can add it to the table.

**Comment by pjaaskel** [comment]
Otherwise LGTM but the CI "random-looking" errors make me worry.

**Review by pjaaskel — APPROVED** [review]

**Pull request merged**